diff -up glib-2.22.5/configurei386-atomic glib-2.22.5/configure
--- glib-2.22.5/configurei386-atomic	2010-03-15 22:52:21.000000000 -0400
+++ glib-2.22.5/configure	2010-03-16 09:54:40.568974681 -0400
@@ -26581,11 +26581,6 @@ $as_echo_n "checking whether to use asse
 
 if test x"$GCC" = xyes; then
   case $host_cpu in
-    i386)
-      { $as_echo "$as_me:${as_lineno-$LINENO}: result: none" >&5
-$as_echo "none" >&6; }
-      glib_memory_barrier_needed=no
-      ;;
     i?86)
       { $as_echo "$as_me:${as_lineno-$LINENO}: result: i486" >&5
 $as_echo "i486" >&6; }
